tests/utils/parse-callstack.py

```diff
@@ -0,0 +1,170 @@
+    recorded_callstack = set()
+    syms=[]
+    addresses=[]
+    # We read kallsyms file and save the output
+    with open('/proc/kallsyms') as kallsyms_f:
+        for line in kallsyms_f:
+            addr = line.split()[0]
+            symbol = line.split()[2]
+            addresses.append(int(addr, 16))
+            syms.append({'addr':int(addr, 16), 'symbol':symbol})
+
+    # Save the address and symbol in a sorted list of tupple
+    syms = sorted(syms, key=lambda k:k['addr'])
+    # We save the list of addresses in a seperate sorted list to easily bisect
+    # the closer address of a symbol.
+    addresses = sorted(addresses)
+
+    # Remove commas and split on spaces
+    for addr in raw_callstack.replace(',', '').split(' '):
+        if '0x' in addr[:2]:
+            # Search the location of the address in the addresses list and
+            # deference this location in the syms list to add the associated
+            # symbol.
+            loc = bisect.bisect(addresses, int(addr, 16))
+            recorded_callstack.add(syms[loc-1]['symbol'])
+
+    return recorded_callstack
+
+# Regex capturing the callstack_user and callstack_kernel context
+user_cs_rexp='.*callstack_user\ \=\ \[(.*)\]\ .*\}, \{.*\}'
+kernel_cs_rexp='.*callstack_kernel\ \=\ \[(.*)\]\ .*\}, \{.*\}'
+
+def main():
+    """
+        Reads a line from stdin and expect it to be a wellformed Babeltrace CLI
+        output containing containing a callstack context of the domain passed
+        as argument.
+    """
+    expected_callstack = set()
+    recorded_callstack = set()
+    cs_type=None
+
+    if len(sys.argv) <= 4:
+        print(sys.argv)
+        raise ValueError('USAGE: ./parse (--kernel|--user EXE) FUNC-NAMES')
+
+    # If the `--user` option is passed, save the next argument as the path
+    # to the executable
+    argc=1
+    executable=None
+    if sys.argv[argc] in '--kernel':
+        rexp = kernel_cs_rexp
+        cs_type='kernel'
+    elif sys.argv[argc] in '--user':
+        rexp = user_cs_rexp
+        cs_type='user'
+        argc+=1
+        executable = sys.argv[argc]
+    else:
+        raise Exception('Unknown domain')
+
+    argc+=1
+
+    # Extract the function names that are expected to be found call stack of
+    # the current events
+    for func in sys.argv[argc:]:
+        expected_callstack.add(func)
+
+    # Read the tested line for STDIN
+    event_line = None
+    for line in sys.stdin:
+        event_line = line
+        break
+
+    # Extract the userspace callstack context of the event
+    m = re.match(rexp, event_line)
+
+    # If there is no match, exit with error
+    if m is None:
+        raise re.error('Callstack not found in event line')
+    else:
+        raw_callstack = str(m.group(1))
+        if cs_type in 'user':
+            recorded_callstack=extract_user_func_names(executable, raw_callstack)
+        elif cs_type in 'kernel':
+            recorded_callstack=extract_kernel_func_names(raw_callstack)
+        else:
```

Comment:
The test fails on my machine. Looking at the resulting traces, I do see that a callstack is captured.
However, the addresses don't match any symbol in `/proc/kallsyms` directly. I don't know if this is because the behaviour of the stack capture code changed at some point.

I'm using `4.16.13-1-ARCH`.

The user callstack part fails too. It seems addr2line can't map the addresses back :confused:.

Reply:
The problem was caused by the fact that on Archlinux (and Ubuntu 18.04) binaries are compiled with the -pie (position independent executable) option making the address resolution more complex. For the test, we turn off this optimization using -no-pie.